feat: Support `meltano init` into an empty directory

ReubenFrankel opened the pull request:
Resolves #6806

### Checklist
- [x] Implement feature
- [x] Add/update tests
- [x] Add/update docs

ReubenFrankel committed:
Support `meltano init` into an empty dir

ReubenFrankel committed:
Add/update tests

ReubenFrankel committed:
Update docs

ReubenFrankel committed:
Merge branch 'main' into 6806-init-into-empty-dir

afolson commented:
@ReubenFrankel Thanks for contributing! I'll make sure someone on the team sees this for review.

github-actions[bot] committed:
chore: Release v2.7.2 (#6825)

tayloramurphy commented on src/meltano/core/project_init_service.py:
```diff
@@ -3,6 +3,7 @@
 
```
I'd request a few more changes to this since it's changing an assumption about the init experience:

* Update the "Created / creating" output

```bash
Created .                   # should be removed
Creating project files...
  ./                        # can this be the current working directory? Or at least the current folder.
   |-- .meltano
   |-- meltano.yml
```

* `Project . has been created!` should be `Your project has been created!`

ReubenFrankel replied:
https://github.com/meltano/meltano/pull/6814/commits/acf4bd22ce3f7d1c665800d84b02b75783ef7577

tayloramurphy commented on src/meltano/core/project_init_service.py:
```diff
@@ -152,8 +154,11 @@ def echo_instructions(self):
+
         click.echo("  Visit ", nl=False)
         click.secho(
             "https://docs.meltano.com/getting-started#create-your-meltano-project",
```
```suggestion
            "https://docs.meltano.com/getting-started/part1",
```

While we're here might as well update this!

aaronsteers committed:
ci: print better version debug info if version mismatch occurs (#6826)

edgarrmondragon commented on tests/meltano/cli/test_initialize.py:
```diff
@@ -53,3 +53,26 @@ def test_init(self, cli_runner, tmp_path_factory, pushd):
         assert "send_anonymous_usage_stats: false" in meltano_yml
         assert "project_id:" in meltano_yml
+
+    def test_init_existing_empty_directory(self, cli_runner, tmp_path_factory, pushd):
```
@ReubenFrankel do you think it makes sense to add a test for the case when the user passes an _absolute_ path. The updated CLI docs say the argument must be a relative directory, so what happens when the user passes an absolute path?

ReubenFrankel replied:
The path can be absolute or relative - I'll update the docs and add a test case. :smile:

ReubenFrankel replied:
https://github.com/meltano/meltano/pull/6814/commits/acf4bd22ce3f7d1c665800d84b02b75783ef7577

ReubenFrankel replied:
https://github.com/meltano/meltano/pull/6814/commits/12d18ee2a1e58b5358f4fc83168b28da812ed02f

dependabot[bot] committed:
chore: bump actions/dependency-review-action from 2.1.0 to 2.4.0

Bumps [actions/dependency-review-action](https://github.com/actions/dependency-review-action) from 2.1.0 to 2.4.0.
- [Release notes](https://github.com/actions/dependency-review-action/releases)
- [Commits](https://github.com/actions/dependency-review-action/compare/v2.1.0...v2.4.0)

---
updated-dependencies:
- dependency-name: actions/dependency-review-action
  dependency-type: direct:production
  update-type: version-update:semver-minor
...

Signed-off-by: dependabot[bot] <support@github.com>

dependabot[bot] committed:
chore: bump actions/setup-node from 3.4.1 to 3.5.0 (#6830)

Bumps [actions/setup-node](https://github.com/actions/setup-node) from 3.4.1 to 3.5.0.
- [Release notes](https://github.com/actions/setup-node/releases)
- [Commits](https://github.com/actions/setup-node/compare/v3.4.1...v3.5.0)

---
updated-dependencies:
- dependency-name: actions/setup-node
  dependency-type: direct:production
  update-type: version-update:semver-minor
...

Signed-off-by: dependabot[bot] <support@github.com>

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>
Co-authored-by: Will Da Silva <will@willdasilva.xyz>

ReubenFrankel committed:
Update URL to getting started guide

Co-authored-by: Taylor A. Murphy <taylor@meltano.com>

ReubenFrankel committed:
Update create progress messages

ReubenFrankel committed:
Test init with relative and absolute paths; correct docs

edgarrmondragon committed:
chore: Enable type checking in `meltano.core.schedule_service` (#6831)

Co-authored-by: Will Da Silva <will@willdasilva.xyz>

dependabot[bot] committed:
chore: bump codecov/codecov-action from 3.1.0 to 3.1.1 (#6828)

Bumps [codecov/codecov-action](https://github.com/codecov/codecov-action) from 3.1.0 to 3.1.1.
- [Release notes](https://github.com/codecov/codecov-action/releases)
- [Changelog](https://github.com/codecov/codecov-action/blob/main/CHANGELOG.md)
- [Commits](https://github.com/codecov/codecov-action/compare/v3.1.0...v3.1.1)

---
updated-dependencies:
- dependency-name: codecov/codecov-action
  dependency-type: direct:production
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>
Co-authored-by: Will Da Silva <will@willdasilva.xyz>

ReubenFrankel committed:
Merge branch 'main' into 6806-init-into-empty-dir

WillDaSilva commented on docs/src/_reference/command-line-interface.md:
```diff
@@ -486,7 +486,7 @@ meltano init [project_name] [--no_usage_stats]
 
 #### Parameters
 
-- **project_name** - This determines the folder name for the project
```
I think `project_name` should be renamed to `project_directory`, both here, in `src/meltano/core/project_init_service.py`, and wherever else the term is used.

WillDaSilva replied:
Aside from that, I think this is good to go. Thanks @ReubenFrankel!

ReubenFrankel replied:
I agree - this did cross my mind but wondered if it was too much of a change initially.

WillDaSilva replied:
The name has always been misleading imo, since Meltano does not store or ever use this "project name" for anything other than creating the directory.

ReubenFrankel replied:
https://github.com/meltano/meltano/pull/6814/commits/1c7a5ec2d9b5ab02671d1bae27c3f8d3b5b7bd18

edgarrmondragon reviewed (approved):
LGTM! I can confirm these 3 modes work fine:

- `meltano init ../my-meltano-project`
- `meltano init /abs/path/to/my-meltano-project`
- `meltano init my-meltano-project`

ReubenFrankel committed:
Small refactor to parse `pathlib.Path` object directly from `meltano init` argument

ReubenFrankel committed:
Update references to project name to project directory

ReubenFrankel committed:
Merge branch '6806-init-into-empty-dir' of github.com:ReubenFrankel/meltano into 6806-init-into-empty-dir

ReubenFrankel committed:
Remove unused method

ReubenFrankel committed:
Disable `WPS210` for `test_init`

ReubenFrankel committed:
Escape exception message match to support Windows path

WillDaSilva commented on src/meltano/core/project_init_service.py:
```diff
@@ -152,23 +149,15 @@ def echo_instructions(self):
+
+        if not Path.cwd().samefile(self.project.root):
+            click.secho("  cd ", nl=False)
+            click.secho(self.project.root, fg="magenta")
```
It should be possible to resolve this by attempting to make `self.project.root` relative to the current working directory using `pathlib`. If it can be made relative, print the relative path (which may just be the name e.g. `testingdir`), and if not, print `self.project.root`.

ReubenFrankel committed:
Show project as created relative to current working directory, if possible

ReubenFrankel committed:
Don't show the project directory path for created files, if in the current working directory

tayloramurphy reviewed (approved):
Appreciate the effort that went into this @ReubenFrankel! Looks great!

edgarrmondragon commented on src/meltano/core/project_init_service.py:
```diff
@@ -91,7 +97,9 @@ def create_files(self, add_discovery=False):
         """
         click.secho("Creating project files...", fg="blue")
-        click.echo(f"  {self.project_name}/")
+
+        if not self.project_directory.samefile(Path.cwd()):
+            click.echo(f"  {self.project_directory}/")
```
`Path.samefile` requires the directory to exist:

```suggestion
        os.makedirs(self.project_directory, exist_ok=True)

        if not self.project_directory.samefile(Path.cwd()):
            click.echo(f"  {self.project_directory}/")
```

ReubenFrankel committed:
Use equality check  over `Path.samefile` between project root and current working directory

ReubenFrankel committed:
Merge branch 'main' into 6806-init-into-empty-dir

WillDaSilva commented on src/meltano/core/project_init_service.py:
```diff
@@ -20,13 +21,19 @@ class ProjectInitServiceError(Exception):
+            project_directory: The directory path to create the project at
         """
-        self.project_name = project_name.lower()
+        self.project = Project(project_directory)
+        self.project_directory = self.project.root
+
+        try:
+            self.project_directory = self.project.root.relative_to(Path.cwd())
+        except ValueError:
+            pass
```
```suggestion
        self.project_directory = Path(project_directory)
```

WillDaSilva commented on src/meltano/core/project_init_service.py:
```diff
@@ -152,23 +155,15 @@ def echo_instructions(self):
-        click.secho("  cd ", nl=False)
-        click.secho(self.project_name, fg="magenta")
+
+        if self.project.root != Path.cwd():
```
```suggestion
        if self.project_directory != Path.cwd():
```

WillDaSilva commented on src/meltano/core/project_init_service.py:
```diff
@@ -42,25 +49,21 @@ def init(self, activate: bool = True, add_discovery: bool = False) -> Project:
-                f"Could not create directory {self.project_name!r}. {ex}"
+                f"Could not create directory '{self.project_directory}'. {ex}"
             ) from ex
 
```
```suggestion

        project = Project(self.project_directory)

```

WillDaSilva commented on src/meltano/core/project_init_service.py:
```diff
@@ -91,7 +94,9 @@ def create_files(self, add_discovery=False):
         """
         click.secho("Creating project files...", fg="blue")
-        click.echo(f"  {self.project_name}/")
+
```
GitHub won't let me put a suggestion above this line, but you'll need to:
- Update the signature of `create_files` to take the `project: Project` parameter, and use it instead of `self.project`.
- Update the signature of `create_dot_meltano_dir` to take the `project: Project` parameter, and use it instead of `self.project`.
- Update the signature of `create_system_database` to take the `project: Project` parameter, and use it instead of `self.project`.
- Update `Project.activate(self.project)` to `Project.activate(project)`.
- Update `return self.project` to `return project`.
- Update `for path in plugin.create_files(self.project):` to `for path in plugin.create_files(project):`.

WillDaSilva commented on src/meltano/core/project_init_service.py:
```diff
@@ -91,7 +94,9 @@ def create_files(self, add_discovery=False):
+        if self.project.root != Path.cwd():
+            click.echo(f"  {self.project_directory}/")
 
         self.create_dot_meltano_dir()
```
```suggestion
        self.create_dot_meltano_dir(project)
```

WillDaSilva commented on src/meltano/core/project_init_service.py:
```diff
@@ -42,25 +49,21 @@ def init(self, activate: bool = True, add_discovery: bool = False) -> Project:
-
-        self.project = Project(self.project_name)
-
         self.create_files(add_discovery=add_discovery)
```
```suggestion
        self.create_files(project, add_discovery=add_discovery)
```

WillDaSilva commented on src/meltano/core/project_init_service.py:
```diff
@@ -42,25 +49,21 @@ def init(self, activate: bool = True, add_discovery: bool = False) -> Project:
-
         self.create_files(add_discovery=add_discovery)
 
         self.settings_service = ProjectSettingsService(self.project)
```
```suggestion
        self.settings_service = ProjectSettingsService(project)
```

WillDaSilva reviewed (changes requested):
Context for my suggested changes: https://meltano.slack.com/archives/C013Z450LCD/p1665516060074119

With these changes applied, the issues on older versions of Python on Windows should be resolved. I've implemented them locally, and it has resolve the issue, but because this PR is from a fork I can't commit them directly. Sorry for the hassle.

ReubenFrankel committed:
Apply suggestions from code review

Co-authored-by: Will Da Silva <will@willdasilva.xyz>

ReubenFrankel committed:
Satisfy tests

ReubenFrankel committed:
Merge branch 'main' into 6806-init-into-empty-dir

WillDaSilva reviewed (approved):
Just tested it out again manually, and didn't find any issues. Thanks @ReubenFrankel!

edgarrmondragon committed:
Merge branch 'main' into 6806-init-into-empty-dir
